### DISCUSSION TITLE
Add support for CanCanCan

### DISCUSSION THREAD

**Discussion by brusacco** [avo-hq/avo#1673] [Feature requests]
Can we add support for CanCanCan ?

**Comment by adrianthedev** [comment]
Yes, you should be able to do that @brusacco 

We recently abstracted away the authorization client and you have instructions on how to do that [here](https://docs.avohq.io/2.0/authorization.html#custom-authorization-clients).
Avo doesn't come with a cancancan adapter, but we'd love to have it if you want to send a PR.

Please check the docs and try to create one. If you hit any snags and you need our support with some changes in Avo or navigating the codebase, please reach out; I'd love to help.

As an approach, I would recommend:

- clone the repo
- duplicate the [`Pundit` client](https://github.com/avo-hq/avo/blob/main/lib/avo/services/authorization_clients/pundit_client.rb) in a `Cancancan` client
- update [this](https://github.com/avo-hq/avo/blob/main/lib/avo/services/authorization_service.rb) to take `:cancancan` as an option
- start integrating Cancancan

**Reply by adrianthedev** [reply]
A more easier route would be to use Pundit policies and instead of adding the logic in there, just forward the requests somehow to an adapter that queries your cancancan.

```ruby
class UserPolicy < ApplicationPolicy
  def index?
    CancancanAdapter :index, user, record
  end

  def show?
    CancancanAdapter :show, user, record
  end
...
```

Of course, you can automate most of that👆, but please read in-depth the docs [authorization page](https://docs.avohq.io/2.0/authorization.html#custom-authorization-clients).
